Fix wrapper to validate HTTP status code before downloading

academey opened the pull request:
Fixes #6557

## Summary
This PR fixes a long-standing issue where the Gradle wrapper would save HTTP error responses (like 404 pages or 302 redirects) as if they were valid ZIP files, causing confusing errors and corrupted cache entries.

## Changes
- Added HTTP status code validation in  after establishing the connection
- The wrapper now fails fast with a clear error message when receiving non-200 status codes
- Prevents corrupted files from being saved to the wrapper cache

## Testing
- Added comprehensive tests for 404 and 302 status code handling
- Tests verify that no files are created when the download fails
- Tests verify appropriate error messages are thrown

## Context
As described in #6557, when a server returns a non-200 status code with a response body (such as an HTML error page), the wrapper would:
1. Save the response body as if it were a valid ZIP file
2. Attempt to unzip this invalid file, causing confusing errors
3. Cache this corrupted file, causing subsequent runs to fail without attempting a new download

This fix ensures that only successful (200 OK) responses are processed and saved.

## Contributor Note
This is my first contribution to Gradle. I've followed the contribution guidelines and would appreciate any feedback on the implementation.

academey commented:
@ov7a I've updated the code based on your feedback
The integration tests are now passing locally. The implementation ensures that
1. HTTP status codes are validated before attempting to download
2. Clear error messages are shown
3. No corrupted files are saved to the wrapper cache on failure

Let me know if there are any other changes needed! Thanks

ov7a commented on platforms/core-runtime/wrapper-main/src/integTest/groovy/org/gradle/integtests/WrapperHttpIntegrationTest.groovy:
```diff
@@ -108,6 +108,26 @@ class WrapperHttpIntegrationTest extends AbstractWrapperIntegrationSpec {
+        !distributionZip().exists()
+    }
+
+    private File distributionZip() {
```
❌ Please move this private method closer to other helpers above and verify that the file is created during the successful download (in `downloads wrapper from http server and caches`)

academey replied:
@ov7a I apologize for the confusion. I've now properly addressed your feedback

The prepareWrapper(String baseUrl) helper method was already correctly positioned with other private methods. And added verifyDistributionDownloaded() helper method to check if wrapper files were downloaded. Updated "downloads wrapper from http server and caches" test to verify the file is created during successful download

All tests in WrapperHttpIntegrationTest are passing.

ov7a replied:
There's no validation that the file was not downloaded on error.

ov7a reviewed (changes requested):
LGTM overall, please fix a minor thing below

ov7a commented:
@bot-gradle test this

ov7a commented:
@academey There are test failures, PTAL and fix them

academey commented:
@ov7a The test failures were due to the expected wrapper JAR hash changing after adding HTTP status validation to Download.java. This is expected behavior when modifying wrapper code.
I've updated the EXPECTED_WRAPPER_JAR_HASH in WrapperGenerationIntegrationTest.groovy from the old value to the new one. All tests are now passing locally, thanks. I'm so sorry for you to review these pr many times

ov7a commented:
@bot-gradle test this

academey commented:
Fixed the CI failures. The issue was with the error message format - ToolingApiRemoteIntegrationTest was expecting the exact format "Server returned HTTP response code: 500 for URL: ..." but our code was producing "Download failed: Server returned HTTP 500 for URL ...".

Reverted to the original error message format to maintain compatibility with existing tests.

ov7a commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":3190898422,"triggeredTeamCityBuilds":[{"buildConfigurationId":"Gradle_Master_Check_Stage_PullRequestFeedback_Trigger","buildId":"101803483","headRef":"d91571663c0855978e2936d32dec273d44824615"}]} -->
I've triggered the following builds:
- [Pull Request Feedback (Trigger) (Check)](https://builds.gradle.org/build/101803483)
  - [Build Scan](https://ge.gradle.org/scans?search.tags=CI&amp;search.names=gitCommitId,PartOf&amp;search.values=d91571663c0855978e2936d32dec273d44824615,PullRequestFeedback&amp;)

ov7a commented:
@academey ignore the above, it's a fluke in CI

academey commented:
@ov7a Thanks for the thorough review and guidance throughout this PR! The integration test approach with BlockingHttpServer is definitely cleaner, and I learned a lot about Gradle's testing infrastructure in the process. I really appreciate your patience thanks!!
